src/Diagrams/Swimunit/Dotmatrix.hs

```diff
@@ -0,0 +1,843 @@
+
+{-|
+  Renders list of characters to a diagram where the characters
+  are shown as if displayed by an old dotmatrix panel.
+  Only the light emitting dots are shown.
+
+  This function maps a list of n characters passed as third argument
+  to n dotmatrix characters.
+ -}
+dotmatrix  :: Diagram B                        -- ^ Diagram to show a 'dot'.
+           -> Dotfont                          -- ^ The Dotfont to be used.
+           -> String                           -- ^ List of characters to be rendered in dotmatrix.
+           -> Diagram B                        -- ^ Resulting diagram.
+dotmatrix  dt fnt txt = dotmatrix' dt fnt False txt
+
+{-|
+  Renders list of characters to a diagram where the characters
+  are shown as if displayed by an old dotmatrix panel.
+  Only the dark dots are shown. This function is usually used to render
+  the back panel.
+
+  Function is calling dotmatrix.
+ -}
+notdotmatrix :: Diagram B                      -- ^ Diagram to show a 'dot'.
+             -> Dotfont                        -- ^ The Dotfont to be used.
+             -> String                         -- ^ List of characters to be rendered in dotmatrix.
+             -> Diagram B                      -- ^ Resulting diagram.
+notdotmatrix dt fnt txt = dotmatrix' dt fnt True txt
+
+{-|
+  Renders list of characters to a diagram where the characters
+  are shown as if displayed by an old dotmatrix panel.
+  Whether light emitting dots or backpanel dots are shown is controlled by third
+  argument.
+
+  This function maps a list of n characters passed as fourth argument
+  to n dotmatrix characters. The third argument decides whether the
+  dots of the font or the dots of the backplane are rendered.
+
+  Devnote: Function loops over list of characters in fourth argument.
+ -}
+dotmatrix' :: Diagram B                       -- ^ Diagram to show a 'dot'.
+           -> Dotfont                         -- ^ The Dotfont to be used.
+           -> Bool                            -- ^ True if inverse of font is returned.
+           -> String                          -- ^ List of characters to be rendered in dotmatrix.
+           -> Diagram B                       -- ^ Resulting diagram.
+dotmatrix' dt fnt rev txt = hcat (map (dotcharblock dt fnt rev) txt)
+
+{-|
+  Vertically stacks a number of lines of dots prepared by function dotline
+  to a block of dots representing a Dotchar.
+ -}
+dotcharblock :: Diagram B                       -- ^ Diagram to show a 'dot'.
+             -> Dotfont                         -- ^ The Dotfont to be used.
+             -> Bool                            -- ^ If true comparison is inverted.
+             -> Char                            -- ^ Character to be mapped to a Dotchar.
+             -> Diagram B                       -- ^ Resulting diagram.
+dotcharblock dt fnt rev c = vcat (map (dotcharline dt rev)
+                                      (dotcharlistoflistsofchars(dotcharmap fnt c))
```

Comment:
Hi Chris,
that is definitely better than my version.
A big thank for your suggestion!

Regards
     Wolfram


On 18.09.2015 21:18, Chris wrote:
>
> In src/Diagrams/Swimunit/Dotmatrix.hs 
> <https://github.com/diagrams/diagrams-contrib/pull/55#discussion_r39891270>:
>
> > +           -> Bool                            -- ^ True if inverse of font is returned.
> > +           -> String                          -- ^ List of characters to be rendered in dotmatrix.
> > +           -> Diagram B                       -- ^ Resulting diagram.
> > +dotmatrix' dt fnt rev txt = hcat (map (dotcharblock dt fnt rev) txt)
> > +
> > +{-|
> > +  Vertically stacks a number of lines of dots prepared by function dotline
> > +  to a block of dots representing a Dotchar.
> > + -}
> > +dotcharblock :: Diagram B                       -- ^ Diagram to show a 'dot'.
> > +             -> Dotfont                         -- ^ The Dotfont to be used.
> > +             -> Bool                            -- ^ If true comparison is inverted.
> > +             -> Char                            -- ^ Character to be mapped to a Dotchar.
> > +             -> Diagram B                       -- ^ Resulting diagram.
> > +dotcharblock dt fnt rev c = vcat (map (dotcharline dt rev)
> > +                                      (dotcharlistoflistsofchars(dotcharmap fnt c))
>
> Sure, they make sense. Personally I'd offer an |invertFont :: Dotfont 
> -> Dotfont| function instead of taking an extra parameter everywhere 
> but it's up to you.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/diagrams/diagrams-contrib/pull/55/files#r39891270>.
>